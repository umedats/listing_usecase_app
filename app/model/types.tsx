import { Subscription as SDKSubscription } from "@/app/sdk/notifications";

/**
 * The Issue returned from the Listing API.
 */
export interface Issue {
  code: string;
  message: string;
  severity: string;
  attributeNames?: string[];
}

/**
 * The result of the Listings Items API submission.
 */
export interface ListingSubmissionResult {
  sku: string;
  status: string;
  submissionId: string;
  issues?: Issue[];
}

/**
 * Relevant information from Listings GET API call.
 */
export interface Listing {
  attributes: object;
  issues?: Issue[];
  productType?: string;
}

/**
 * A type which represents a patch for the contribution.
 */
export interface Patch {
  op: string;
  path: string;
  value: any;
}

/**
 * A type for the alert dialog state.
 */
export interface AlertDialogState {
  showAlert: boolean;
  alertTitle: string;
  alertContent: string;
}

/**
 * A type which captures the SP API request and response.
 */
export interface SPAPIRequestResponse {
  apiName: string;
  apiDocumentationLink: string;
  request: object;
  response: object;
}

/**
 * A type which capture the information about a feed.
 */
export interface Feed {
  feedId: string;
  feedType: string;
  createdTime: string;
  processingStatus: string;
  resultFeedDocumentId: string;
}

/**
 * A type which capture the information about a Subscription.
 */
export interface Subscription {
  notificationType: string;
  eventBusName?: string;
  eventBusFilterRule?: string;
  sqsTargetId?: string;
  sqsQueueName?: string;
  subscriptionId: string;
}

/**
 * A type which captures the raw SDK subscription along with the
 * notification type.
 */
export interface RawSubscription {
  notificationType: string;
  subscription: SDKSubscription;
}
