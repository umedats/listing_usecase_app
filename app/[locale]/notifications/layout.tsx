"use client";
import React from "react";
import DualPageLayout from "@/app/components/dual-page-layout";

/**
 * Parallel Route layout for the Notifications.
 * @param props input props.
 * @constructor
 */
export default function Layout(props: {
  createSubscription: React.ReactNode;
  viewSubscriptions: React.ReactNode;
}) {
  return (
    <DualPageLayout
      leftContent={props.createSubscription}
      rightContent={props.viewSubscriptions}
    />
  );
}
