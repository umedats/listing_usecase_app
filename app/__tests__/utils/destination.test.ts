import getSPAPIEndpoint, {
  buildGrantLessNotificationsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import { NotificationsApi } from "@/app/sdk/notifications";
import { SPAPIRequestResponse } from "@/app/model/types";
import {
  createEventBridgeDestination,
  getEventBridgeDestinations,
} from "@/app/utils/destination";

jest.mock("@/app/sdk/factory/ListingsSPAPIClientsFactory");
const mockedGetSPAPIEndpoint = jest.mocked(getSPAPIEndpoint);
mockedGetSPAPIEndpoint.mockReturnValue("");

const notificationsApi = new NotificationsApi(undefined);
const mockedGetDestinations = jest.fn();
notificationsApi.getDestinations = mockedGetDestinations;
const mockedCreateDestinations = jest.fn();
notificationsApi.createDestination = mockedCreateDestinations;

const mockedBuildGrantLessNotificationsAPIClient = jest.mocked(
  buildGrantLessNotificationsAPIClient,
);
mockedBuildGrantLessNotificationsAPIClient.mockResolvedValue(notificationsApi);

describe("Test for all the SP-API Destination APIs", () => {
  describe("Test for the getEventBridgeDestinations", () => {
    test("snapshot test for 200 response from GetDestinations API", async () => {
      mockedGetDestinations.mockResolvedValueOnce({
        payload: [
          {
            name: "TestDestination",
            destinationId: "5fe08e61-c5ec-4427-9f25-24a7b630cce2",
            resource: {
              eventBridge: {
                name: "sellingpartnerapi.amazon.com/amzn1.sellerapps.app.15a75829-cd4a-4efc-b947-0cc39d874577",
                region: "us-east-1",
                accountId: "123456",
              },
            },
          },
        ],
      });

      const reqResponses: SPAPIRequestResponse[] = [];
      const destinations = await getEventBridgeDestinations(
        "us-east-1",
        "123456",
        reqResponses,
      );
      expect(destinations).toMatchSnapshot();
      expect(reqResponses).toMatchSnapshot();
    });

    test("snapshot test for the 404 response from GetDestinations API", async () => {
      mockedGetDestinations.mockRejectedValue({
        status: 404,
      });

      const reqResponses: SPAPIRequestResponse[] = [];
      const destinations = await getEventBridgeDestinations(
        "us-east-1",
        "123456",
        reqResponses,
      );
      expect(destinations).toMatchSnapshot();
      expect(reqResponses).toMatchSnapshot();
    });

    test("snapshot test for the 500 response from GetDestinations API", async () => {
      mockedGetDestinations.mockRejectedValue({
        status: 500,
      });

      const reqResponses: SPAPIRequestResponse[] = [];
      await expect(
        getEventBridgeDestinations("us-east-1", "123456", reqResponses),
      ).rejects.toMatchObject({
        status: 500,
      });
      expect(reqResponses).toMatchSnapshot();
    });
  });

  describe("Test for the createDestinations SP-API", () => {
    test("snapshot test for 200 response from CreateDestinations API", async () => {
      mockedCreateDestinations.mockResolvedValueOnce({
        resourceSpecification: {
          eventBridge: {
            region: "us-east-1",
            accountId: "123456",
          },
        },
        name: "Custom Destination",
      });
      const reqResponses: SPAPIRequestResponse[] = [];
      const createDestinationResponse = await createEventBridgeDestination(
        "us-east-1",
        "123456",
        reqResponses,
      );

      expect(createDestinationResponse).toMatchSnapshot();
      expect(reqResponses).toMatchSnapshot();
    });

    test("snapshot test for the 404 response from CreateDestinations API", async () => {
      mockedCreateDestinations.mockRejectedValue({
        status: 404,
      });

      const reqResponses: SPAPIRequestResponse[] = [];
      await expect(
        createEventBridgeDestination("us-east-1", "123456", reqResponses),
      ).rejects.toMatchObject({
        status: 404,
      });
      expect(reqResponses).toMatchSnapshot();
    });
  });
});
