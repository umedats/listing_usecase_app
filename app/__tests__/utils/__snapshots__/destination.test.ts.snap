// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for all the SP-API Destination APIs Test for the createDestinations SP-API snapshot test for 200 response from CreateDestinations API 1`] = `
{
  "name": "Custom Destination",
  "resourceSpecification": {
    "eventBridge": {
      "accountId": "123456",
      "region": "us-east-1",
    },
  },
}
`;

exports[`Test for all the SP-API Destination APIs Test for the createDestinations SP-API snapshot test for 200 response from CreateDestinations API 2`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#post-notificationsv1destinations",
    "apiName": "CreateDestinations",
    "request": {
      "body": CreateDestinationRequest {
        "name": "EventBridgeDestination",
        "resourceSpecification": DestinationResourceSpecification {
          "eventBridge": EventBridgeResourceSpecification {
            "accountId": "us-east-1",
            "region": "123456",
          },
        },
      },
    },
    "response": {
      "name": "Custom Destination",
      "resourceSpecification": {
        "eventBridge": {
          "accountId": "123456",
          "region": "us-east-1",
        },
      },
    },
  },
]
`;

exports[`Test for all the SP-API Destination APIs Test for the createDestinations SP-API snapshot test for the 404 response from CreateDestinations API 1`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#post-notificationsv1destinations",
    "apiName": "CreateDestinations",
    "request": {
      "body": CreateDestinationRequest {
        "name": "EventBridgeDestination",
        "resourceSpecification": DestinationResourceSpecification {
          "eventBridge": EventBridgeResourceSpecification {
            "accountId": "us-east-1",
            "region": "123456",
          },
        },
      },
    },
    "response": {
      "status": 404,
    },
  },
]
`;

exports[`Test for all the SP-API Destination APIs Test for the getEventBridgeDestinations snapshot test for 200 response from GetDestinations API 1`] = `
DestinationList [
  {
    "destinationId": "5fe08e61-c5ec-4427-9f25-24a7b630cce2",
    "name": "TestDestination",
    "resource": {
      "eventBridge": {
        "accountId": "123456",
        "name": "sellingpartnerapi.amazon.com/amzn1.sellerapps.app.15a75829-cd4a-4efc-b947-0cc39d874577",
        "region": "us-east-1",
      },
    },
  },
]
`;

exports[`Test for all the SP-API Destination APIs Test for the getEventBridgeDestinations snapshot test for 200 response from GetDestinations API 2`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#getdestinations",
    "apiName": "GetDestinations",
    "request": {},
    "response": GetDestinationsResponse {
      "payload": DestinationList [
        {
          "destinationId": "5fe08e61-c5ec-4427-9f25-24a7b630cce2",
          "name": "TestDestination",
          "resource": {
            "eventBridge": {
              "accountId": "123456",
              "name": "sellingpartnerapi.amazon.com/amzn1.sellerapps.app.15a75829-cd4a-4efc-b947-0cc39d874577",
              "region": "us-east-1",
            },
          },
        },
      ],
    },
  },
]
`;

exports[`Test for all the SP-API Destination APIs Test for the getEventBridgeDestinations snapshot test for the 404 response from GetDestinations API 1`] = `undefined`;

exports[`Test for all the SP-API Destination APIs Test for the getEventBridgeDestinations snapshot test for the 404 response from GetDestinations API 2`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#getdestinations",
    "apiName": "GetDestinations",
    "request": {},
    "response": {
      "status": 404,
    },
  },
]
`;

exports[`Test for all the SP-API Destination APIs Test for the getEventBridgeDestinations snapshot test for the 500 response from GetDestinations API 1`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#getdestinations",
    "apiName": "GetDestinations",
    "request": {},
    "response": {
      "status": 500,
    },
  },
]
`;
